code/game/objects/items/devices/uplinks.dm

```diff
@@ -233,6 +292,70 @@ GLOBAL_LIST_EMPTY(world_uplinks)
 			var/datum/uplink_item/UI = uplink_items[params["item"]]
 			return buy(UI, UI ? UI.reference : "")
 
+		if("add_to_cart")
+			var/datum/uplink_item/UI = uplink_items[params["item"]]
+			if(!LAZYIN(shopping_cart, item_reference))
+				var/startamount = 1
+				if(UI.limited_stock == 0)
+					startamount = 0
+				LAZYSET(shopping_cart, item_reference, startamount)
+				generate_tgui_cart(TRUE)
+			else
+				to_chat(ui.user, "<span class='warning'>[UI.name] is already in your cart!</span>")
+
+		if("remove_from_cart")
+			remove_from_cart(item_reference)
+
+		if("set_cart_item_quantity")
+			LAZYSET(shopping_cart, item_reference, text2num(params["quantity"]))
+			generate_tgui_cart(TRUE)
+
+		if("purchase_cart")
+			if(!LAZYLEN(shopping_cart))
+				return
+			if(calculate_cart_tc() > uses)
+				to_chat(usr, "<span class='warning'>[src] buzzes, it doesn't contain enough telecrystals!</span>")
```

Comment:
You're inconsistent with usr and UI.user in this proc, just pick one

Reply:
to be fair, i never typed usr (i just love copying code)